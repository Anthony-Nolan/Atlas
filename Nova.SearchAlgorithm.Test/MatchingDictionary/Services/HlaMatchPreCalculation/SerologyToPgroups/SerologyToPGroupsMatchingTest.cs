using ApprovalTests;
using ApprovalTests.Reporters;
using Nova.SearchAlgorithm.MatchingDictionary.Models.HLATypings;
using Nova.SearchAlgorithm.MatchingDictionary.Models.MatchingTypings;
using NUnit.Framework;
using System.Collections.Generic;
using System.Linq;

namespace Nova.SearchAlgorithm.Test.MatchingDictionary.Services.HlaMatchPreCalculation.SerologyToPgroups
{
    [UseReporter(typeof(NUnitReporter))]
    public class SerologyToPGroupsMatchingTest : MatchedOnTestBase<MatchedSerology>
    {
        [Test]
        public void ValidSerologyHaveAtLeastOnePGroup()
        {
            var pGroupCounts = MatchedHla
                .Where(m => !m.HlaTyping.IsDeleted && m.HlaTyping is SerologyTyping)
                .Select(m => new { HlaType = m.HlaTyping, PGroupCount = m.MatchingPGroups.Count() })
                .ToList();

            Assert.IsTrue(pGroupCounts.All(p => p.PGroupCount > 0));
        }

        [Test]
        public void BroadWhereSplitHasAssociatedPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString(MatchLocus.A, "9"));
        }

        [Test]
        public void BroadWhereSplitHasNoAssociatedPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString(MatchLocus.Dqb1, "1"));
        }

        [Test]
        public void BroadHasSplitAndAssociatedPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString(MatchLocus.B, "21"));
        }

        [Test]
        public void SplitHasAssociatedPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString(MatchLocus.B, "51"));
        }

        [Test]
        public void SplitNoAssociatedPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString(MatchLocus.C, "10"));
        }

        [Test]
        public void AssociatedWithSplitPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString(MatchLocus.B, "3902"));
        }

        [Test]
        public void AssociatedWithBroadPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString(MatchLocus.B, "4005"));
        }

        [Test]
        public void AssociatedWithNotSplitPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString(MatchLocus.Drb1, "103"));
        }

        [Test]
        public void NotSplitWithAssociatedPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString(MatchLocus.A, "2"));
        }

        [Test]
        public void NotSplitNoAssociatedPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString(MatchLocus.A, "1"));
        }

        [Test]
        public void DeletedPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString(MatchLocus.C, "11"));
        }

        [Test]
        public void B15BroadPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString(MatchLocus.B, "15"));
        }

        [Test]
        public void B15SplitPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString(MatchLocus.B, "62"));
        }

        [Test]
        public void B70BroadPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString(MatchLocus.B, "70"));
        }

        [Test]
        public void B70SplitPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString(MatchLocus.B, "72"));
        }

        [Test]
        public void B15And70ShareMatchingPGroups()
        {
            var b15 = GetSingleMatchingTyping(MatchLocus.B, "15").MatchingPGroups;
            var b70 = GetSingleMatchingTyping(MatchLocus.B, "70").MatchingPGroups;

            Approvals.Verify(string.Join("\r\n", b15.Intersect(b70).OrderBy(p => p)));
        }

        private string GetPGroupsAsString(MatchLocus matchLocus, string serologyName)
        {
            return string.Join("\r\n",
                GetSingleMatchingTyping(matchLocus, serologyName)
                .MatchingPGroups
                .OrderBy(p => p));
        }
    }
}
