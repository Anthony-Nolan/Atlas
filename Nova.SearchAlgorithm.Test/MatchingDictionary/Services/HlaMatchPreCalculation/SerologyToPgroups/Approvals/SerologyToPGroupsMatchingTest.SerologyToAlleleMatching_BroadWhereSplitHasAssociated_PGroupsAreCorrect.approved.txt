23:01P
23:02
23:03P
23:04
23:05
23:06
23:09
23:10
23:12
23:13
23:14P
23:15
23:16
23:21
23:22
23:23
23:24
23:25
23:26
23:27
23:28
23:29
23:30
23:31
23:32
23:33
23:34
23:35
23:36
23:37P
23:39
23:40
23:41
23:42
23:43
23:44
23:45
23:46
23:47
23:48
23:49
23:50
23:51
23:52
23:53
23:54
23:55
23:56
23:57
23:59
23:60
23:61
23:62
23:63
23:64
23:65
23:66
23:67
23:68
23:70
23:71
23:72
23:73
23:74
23:75
23:76
23:77
23:78
23:79
23:80
23:81
23:82
23:83
24:02P
24:03P
24:04
24:05P
24:06
24:07P
24:08
24:100
24:101
24:102
24:103
24:104
24:105
24:106
24:107
24:108
24:109
24:10P
24:110
24:111
24:112
24:113
24:114
24:115
24:116
24:117
24:118
24:119
24:120
24:121
24:122
24:123
24:124
24:125
24:126
24:127
24:128
24:129
24:130
24:131
24:133
24:134
24:135P
24:136
24:137
24:138
24:139
24:13P
24:140
24:141
24:142P
24:143
24:145
24:146
24:147
24:148
24:149
24:14P
24:15
24:151
24:152
24:156
24:157
24:159
24:160
24:161
24:162
24:164
24:165
24:166
24:167
24:168
24:169
24:17
24:170
24:171
24:172P
24:173
24:174
24:175
24:176
24:177
24:178
24:179
24:18
24:180
24:181
24:182
24:184
24:186
24:187
24:188
24:189
24:19
24:190
24:191
24:192
24:193
24:194
24:195
24:196
24:197
24:198
24:200
24:201
24:202
24:203
24:204
24:205
24:206
24:207P
24:208
24:209
24:20P
24:210
24:212
24:213
24:214
24:215
24:216
24:217
24:218
24:219
24:21P
24:22
24:220
24:221
24:223
24:224
24:225P
24:226P
24:227
24:228
24:229
24:23
24:230
24:233
24:234
24:235
24:236
24:237
24:238
24:239
24:24
24:241
24:242
24:243
24:244
24:245
24:246
24:247
24:248
24:25
24:253
24:254
24:255
24:256
24:257
24:258
24:259
24:260
24:261
24:262
24:26P
24:27
24:272
24:273
24:274
24:275
24:276
24:277
24:279
24:28
24:280
24:281
24:282
24:283
24:284
24:285
24:286
24:287
24:288
24:289
24:29
24:290
24:291
24:292
24:293
24:294Q
24:295
24:296
24:297
24:298
24:299
24:30
24:300
24:301
24:302
24:304
24:305
24:306
24:307
24:308
24:309
24:31
24:310P
24:311
24:313P
24:315
24:316
24:317
24:318
24:319
24:32
24:320
24:321
24:322
24:324
24:325
24:326
24:327
24:328
24:329
24:330
24:331
24:332
24:333
24:334
24:335
24:336
24:337
24:338
24:339
24:34
24:340
24:341
24:342
24:343
24:344
24:345
24:346
24:347P
24:348
24:349
24:35
24:350
24:351
24:355
24:356
24:358
24:360
24:361
24:362
24:363
24:364
24:365
24:366
24:367
24:368
24:369
24:37
24:371
24:372
24:373
24:374
24:375
24:376
24:377
24:378
24:379
24:38
24:380
24:381
24:382
24:384
24:386
24:387
24:39
24:390
24:391
24:392
24:393
24:394
24:395
24:397
24:398
24:399
24:403
24:404
24:405
24:41
24:42
24:43
24:44
24:46
24:47
24:49
24:50
24:51
24:52
24:53
24:54
24:55
24:56
24:57
24:58
24:59
24:61
24:62
24:63
24:64
24:66
24:67
24:68
24:69
24:70
24:71
24:72
24:73
24:74P
24:75
24:77
24:78
24:80
24:81
24:82
24:85
24:87
24:88
24:89
24:91
24:92
24:93
24:94
24:95
24:96
24:97
24:98
24:99
31:08